File: Matrix/matrix.cpp
#include <iostream>
#include "matrix.h"
#include <math.h>

template class Matrix<double>;

template<class TYPE>
Matrix<TYPE>::Matrix()
{

}

template<class TYPE>
Matrix<TYPE>::~Matrix()
{
  for (int i = 0; i < size[0]; i++)
  {
    delete mat[i];
  }
  delete mat;

  delete size;
}

template<class TYPE>
Matrix<TYPE>::Matrix(TYPE** a, int row, int col)
{
  this->mat = new TYPE*[row];
  for (int i = 0; i < row; i++)
  {
    this->mat[i] = new TYPE[col];
    for (int j = 0; j < col; j++)
    {
      this->mat[i][j] = a[i][j];
    }
  }
  this->size = new int(2);
  size[0] = row;
  size[1] = col;
}

template<class TYPE>
Matrix<TYPE>::Matrix(TYPE *a, int size)
{
    this->mat = new TYPE*[1];
    for (int i = 0; i < 1; i++)
    {
        this->mat[i] = new TYPE[size];
        for (int j = 0; j < size; j++)
        {
            this->mat[i][j] = a[j];
        }
    }
    this->size = new int[2];
    this->size[0] = 1;
    this->size[1] = size;
}

template<class TYPE>
void Matrix<TYPE>::printMatrix() const
{
  for (int i = 0; i < size[0]; i++)
  {
    for (int j = 0; j < size[1]; j++)
    {
      std::cout<< mat[i][j]<< "\t";
    }
    std::cout << '\n';
  }
}

template<class TYPE>
void Matrix<TYPE>::printRow(int num) const
{
  if (num >= size[0] || num < 0)
  {
    std::cout << "Out of bounds";
    return;
  }

  for (int i = 0; i < size[1]; i++)
  {
    std::cout << mat[num][i] << '\t';
  }
  std::cout << '\n';
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::sum(Matrix* b)
{
  if (this->size[0] != b->getSize()[0] || this->size[1] != b->getSize()[1])
    return 0;

  TYPE** newMatrix = new TYPE*[this->size[0]];
  for (int i = 0; i < this->size[0]; i++)
  {
    newMatrix[i] = new TYPE[this->size[1]];
    for (int j = 0; j < this->size[1]; j++)
    {
      newMatrix[i][j] = this->getMatrix()[i][j] + b->getMatrix()[i][j];
    }
  }

  return new Matrix<TYPE>(newMatrix, this->size[0], this->size[1]);
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::product(Matrix* b)
{
  int* newSize = new int(2);
  newSize[0] = this->getSize()[0];
  newSize[1] = b->getSize()[1];

  TYPE** newMatrix = new TYPE*[newSize[0]];
  for(int i = 0; i < newSize[0]; i++)
  {
    newMatrix[i] = new TYPE[newSize[1]];
    for(int j = 0; j < newSize[1]; j++)
    {
      TYPE sum = 0;
      for (int k = 0; k < b->getSize()[0]; k++)
      {
        sum += this->getMatrix()[i][k] * b->getMatrix()[k][j];
      }
      newMatrix[i][j] = sum;
      std::cout << newMatrix[i][j] << '\t';
    }
    std::cout << '\n';
  }

  return new Matrix(newMatrix, newSize[0], newSize[1]);
}

//TODO: доработай эту дрисню
//      я знаю тебе хочется это стереть, но чувак, ты должен её доделать,
//      проблема только с блоками ссаными

//UPD: done! проверка для 2 одинаковых матриц 2х2 и
//     для матриц 2х2 и 3х3
//TODO: проверка для неквадратных матриц:
//UPD: не работает для неквадратных (скорее всего и не надо)
template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::kroneckerProduct(Matrix* b)
{
  int* newSize = new int[2];
  newSize[0] = this->getSize()[0] * b->getSize()[0];
  newSize[1] = this->getSize()[1] * b->getSize()[1];

  int nBlockMatrix = this->getSize()[0] * this->getSize()[1];

  TYPE*** blockMatrix = new TYPE**[nBlockMatrix];
  for (int i = 0; i < nBlockMatrix; i++)
  {
    blockMatrix[i] = new TYPE*[b->getSize()[0]];
    for (int j = 0; j < b->getSize()[0]; j++)
    {
      blockMatrix[i][j] = new TYPE[b->getSize()[1]];
    }
  }


  int block = 0;
  for (int l = 0; l < nBlockMatrix/this->getSize()[0]; l++)
  {
    for (int k = 0; k < nBlockMatrix/this->getSize()[0]; k++)
    {
      for (int i = 0; i < b->getSize()[0]; i++)
      {
        for (int j = 0; j < b->getSize()[1]; j++)
        {
          blockMatrix[block][i][j] = this->getMatrix()[l][k] * b->getMatrix()[i][j];
        }
      }
      block++;
    }
  }

  std::cout << "\nNow matrixes go\n";

  for (int k = 0; k < nBlockMatrix; k++)
  {
      for (int i = 0; i < b->getSize()[0]; i++)
      {
          for (int j = 0; j < b->getSize()[1]; j++)
          {
              std::cout << blockMatrix[k][i][j] << "\t";
          }
          std::cout << '\n';
      }
      std::cout << "\n\n";
  }

  TYPE** newMatrix = new TYPE*[newSize[0]];
  for (int i = 0; i < newSize[0]; i++)
  {
    newMatrix[i] = new TYPE[newSize[1]];
  }



  int l = 0;
  int k = 0;
  block = 0;


  for (int i = 0; i < newSize[0]; i++)
  {
      if (k >= b->getSize()[0] && i != newSize[0] - 1)
      {
          k = 0;
          block += 2;
      }

      l = 0;
      for (int j = 0; j < newSize[1]; j++)
      {
          if (l >= b->getSize()[1] && j != newSize[1] - 1)
          {
              l = 0;
              block++;
          }
          newMatrix[i][j] = blockMatrix[block][k][l];
          l++;
      }
      block--;
      k++;
  }



  return new Matrix(newMatrix, newSize[0], newSize[1]);
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::shultsReverse()
{
    int* size = new int[2];
    size[0] = getSize()[0];
    size[1] = getSize()[1];
    double** edinMat = new double*[size[0]];
    for (int i = 0; i < size[0]; i++)
    {
        edinMat[i] = new double[size[1]];
        for (int j = 0; j < size[1]; j++)
        {
            edinMat[i][j] = 0;
        }

        edinMat[i][i] = 2;
    }

    Matrix<double>* edin = new Matrix<double>(edinMat, size[0], size[1]);

    double N1 = 0, Ninf = 0; //норма матрицы по столбцам и по строкам
    Matrix<double>* inv = clone(this);
    inv->printMatrix();
    /*
    for (int row = 0; row < size[0]; row++){
        double colsum = 0, rowsum = 0;
        for(int col = 0; col < size[1]; col++){
            rowsum += fabs(inv->getMatrix()[row][col]);
            colsum += fabs(inv->getMatrix()[col][row]);
        }
        N1 = std::max(colsum, N1);
        Ninf = std::max(rowsum, Ninf);
    }

    inv = inv->transpose()->productNumber(1/(N1*Ninf));
    double EPS = 0.001;

    while (fabs (this->product(inv)->getGaussDet() - 1) >= EPS)
    {
        Matrix* prev = clone(inv);
        inv = this->product(prev);
        inv = inv->productNumber(-1);
        inv = inv->sum(edin);
        inv = prev->product(inv);
        delete prev;
    }
    delete edinMat;
    */
    return inv;
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::transpose()
{
    TYPE** trasposedMat = new TYPE *[this->getSize()[1]];
    for (int i = 0; i < this->getSize()[1]; i++)
    {
        trasposedMat[i] = new TYPE[this->getSize()[0]];
        for (int j = 0; j < this->getSize()[0]; j++)
        {
            trasposedMat[i][j] = this->getMatrix()[j][i];
        }
    }

    return new Matrix(trasposedMat, this->getSize()[0], this->getSize()[1]);
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::productNumber(double num)
{
    for (int i = 0; i < this->getSize()[0]; i++)
    {
        for (int j = 0; j < this->getSize()[1]; j++)
        {
            this->getMatrix()[i][j] *= num;
        }
    }

    return new Matrix(this->getMatrix(), this->getSize()[0], this->getSize()[1]);
}

template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::minus()
{
    double** a = new double*[this->getSize()[0]];
    for (int i = 0; i < this->getSize()[0]; i++)
    {
        a[i] = new double[this->getSize()[1]];
        for (int j = 0; j < this->getSize()[1]; j++)
        {
            a[i][j] = -this->getMatrix()[i][j];
        }
    }

    return new Matrix(a, this->getSize()[0], this->getSize()[1]);
}


template<class TYPE>
Matrix<TYPE>* Matrix<TYPE>::clone(Matrix *a)
{
    TYPE** newMat = new TYPE*[a->getSize()[0]];
    for (int i = 0; i < a->getSize()[0]; i++)
    {
        newMat[i] = new TYPE[a->getSize()[1]];
        for (int j = 0; j < a->getSize()[1]; j++)
        {
            newMat[i][j] = a->getMatrix()[i][j];
        }
    }

    return new Matrix(newMat, a->getSize()[0], a->getSize()[1]);
}

template<class TYPE>
double Matrix<TYPE>::getGaussDet()
{
    double det = 1;
    for (int i = 0; i < this->getSize()[0]; i++)
    {
        det *= this->getMatrix()[i][i];
    }
    return det;
}

template<class TYPE>
int* Matrix<TYPE>::getSize()
{
  return size;
}

template<class TYPE>
TYPE** Matrix<TYPE>::getMatrix()
{
  return mat;
}

File: c#/MatrixLib/MatrixLib/Matrix.cs
using System;

namespace MatrixLib
{
    public class Matrix
    {
        #region Vars
        public double[,] matrix { get; }
        private int row, col;
        #endregion

        private const double EPS = 0.001;

        #region Constructions
        public Matrix()
        {

        }

        public Matrix(double[,] mat, int row, int col)
        {
            this.matrix = mat;
            this.row = row;
            this.col = col;
        }

        public Matrix(double[] vec, int size)
        {
            double[,] mat = new double[1, size];

            for (int i = 0; i < size; i++)
            {
                mat[0, i] = vec[i];
            }

            this.matrix = mat;
            this.row = 1;
            this.col = size;
        }
        #endregion

        public void printMatrix()
        {
            for (int i = 0; i < row; i++)
            {
                for (int j = 0; j < col; j++)
                {
                    Console.Write(matrix[i, j] + "\t");
                }
                Console.WriteLine();
            }
        }

        public static Matrix operator +(Matrix a, Matrix b)
        {
            double[,] newMat = new double[a.row, a.col];
            for (int i = 0; i < a.row; i++)
            {
                for (int j = 0; j < a.col; j++)
                {
                    newMat[i, j] = a.matrix[i, j] + b.matrix[i, j];
                }
            }

            return new Matrix(newMat, a.row, a.col);
        }

        public static Matrix operator -(Matrix a, Matrix b)
        {
            double[,] newMat = new double[a.row, a.col];
            for (int i = 0; i < a.row; i++)
            {
                for (int j = 0; j < a.col; j++)
                {
                    newMat[i, j] = a.matrix[i, j] - b.matrix[i, j];
                }
            }

            return new Matrix(newMat, a.row, a.col);
        }

        public static Matrix operator *(Matrix a, Matrix b)
        {
            double[,] newMat = new double[a.row, a.col];

            for (int i = 0; i < a.row; i++)
            {
                for (int j = 0; j < a.col; j++)
                {
                    newMat[i, j] = 0;
                    for (int k = 0; k < a.row; k++)
                    {
                        newMat[i, j] += a.matrix[i, k] * b.matrix[k, j];
                    }
                    if (Math.Abs(newMat[i, j]) < EPS)
                        newMat[i, j] = 0;
                }
            }

            return new Matrix(newMat, a.row, a.col);
        }

        public Matrix GetMultiplication(Matrix a, Matrix b)
        {
            double[,] newMat = new double[a.row, a.col];

            for (int i = 0; i < a.row; i++)
            {
                for (int j = 0; j < a.col; j++)
                {
                    newMat[i, j] = 0;
                    for (int k = 0; k < a.row; k++)
                    {
                        newMat[i, j] += a.matrix[i, k] * b.matrix[k, j];
                    }
                    if (Math.Abs(newMat[i, j]) < EPS)
                        newMat[i, j] = 0;
                }
            }

            return new Matrix(newMat, a.row, a.col);
        }

        public Matrix kroneckerProduct(Matrix b)
        {
            int tempRow = this.row * b.row;
            int tempCol = this.col * b.col;

            int nBlocks = this.row * b.col;

            double[,,] blockMatrix = new double[nBlocks, b.row, this.col];

            int k, l;
            int block = 0;
            for (l = 0; l < nBlocks / this.row; l++)
            {
                for (k = 0; k < nBlocks / this.row; k++)
                {
                    for (int i = 0; i < b.row; i++)
                    {
                        for (int j = 0; j < b.col; j++)
                        {
                            blockMatrix[block, i, j] = this.matrix[l, k] * b.matrix[i, j];
                        }
                    }
                    block++;
                }
            }

            double[,] tempMatrix = new double[tempRow, tempCol];

            l = 0;
            k = 0;
            block = 0;
            for (int i = 0; i < tempRow; i++)
            {
                if (k >= b.row && i != tempRow - 1)
                {
                    k = 0;
                    block += 2;
                }

                l = 0;
                for (int j = 0; j < tempCol; j++)
                {
                    if (l >= b.col && j != tempCol - 1)
                    {
                        l = 0;
                        block++;
                    }
                    tempMatrix[i, j] = blockMatrix[block, k, l];
                    l++;
                }
                block--;
                k++;
            }

            return new Matrix(tempMatrix, tempRow, tempCol);
        }

        public Matrix invert()
        {
            Matrix mat = clone();
            Matrix edin = Matrix.GetE2Matrix(row);

            double N1 = 0, NInf = 0;
            Matrix inv = clone();


            for (int i = 0; i < row; i++)
            {
                double colsum = 0, rowsum = 0;
                for (int j = 0; j < col; j++)
                {
                    rowsum += Math.Abs(inv.matrix[i, j]);
                    colsum += Math.Abs(inv.matrix[j, i]);
                }
                N1 = Math.Max(colsum, N1);
                NInf = Math.Max(rowsum, NInf);
            }

            inv = inv.transpose();
            inv = inv.productNumber(1 / (N1 * NInf));
            
            while (Math.Abs((mat*inv).getGaussDet() - 1) >= EPS)
            {
                Matrix prev = inv.clone();
                inv = mat * prev;
                inv = inv.productNumber(-1);
                inv = inv + edin;
                inv = prev * inv;
            }

            return inv;
        }

        public static Matrix GetE2Matrix(int size)
        {
            double[,] edin = new double[size, size];
            for(int i = 0; i < size; i++)
            {
                for (int j = 0; j < size; j++)
                {
                    if (i == j)
                        edin[i, j] = 2;
                    else
                        edin[i, j] = 0;
                }
            }

            return new Matrix(edin, size, size);
        }

        public Matrix transpose()
        {
            double[,] tempMatrix = new double[this.col, this.row];
            for (int i = 0; i < this.col; i++)
            {
                for (int j = 0; j < this.row; j++)
                {
                    tempMatrix[i, j] = this.matrix[j, i];
                }
            }
            return new Matrix(tempMatrix, this.col, this.row);
        }

        public Matrix productNumber(double num)
        {
            double[,] tempMatrix = this.matrix;
            for (int i = 0; i < this.row; i++)
            {
                for (int j = 0; j < this.col; j++)
                {
                    tempMatrix[i, j] *= num;
                }
            }
            return new Matrix(tempMatrix, this.row, this.col);
        }

        public double getGaussDet()
        {
            double det = 1;
            for (int i = 0; i < this.row; i++)
            {
                det *= this.matrix[i, i];
            }
            return det;
        }

        public Matrix clone()
        {
            double[,] mat = new double[row, col];

            for (int i = 0; i < row; i++)
            {
                for (int j = 0; j < col; j++)
                {
                    mat[i, j] = matrix[i, j];
                }
            }

            return new Matrix(mat, row, col);
        }
    }
}
